src/main/java/com/spaceuptech/space_api/utils/Service.java

```diff
@@ -0,0 +1,136 @@
+                        finishedLatch.countDown();
+                    }
+                }
+            } else if (payload.getType().equals(Constants.TYPE_SERVICE_REQUEST)) {
+                if (methods.containsKey(payload.getFunction())) {
+                    try {
+                        methods.get(payload.getFunction()).onInvocation(new Message(payload.getParams()), new Message(payload.getAuth()), new ReturnCallback() {
+                            @Override
+                            public void send(String type, Object answer) {
+                                if (type.equals("response")) {
+                                    sendStreamRef.get().onNext(FunctionsPayload.newBuilder()
+                                            .setId(payload.getId())
+                                            .setType(Constants.TYPE_SERVICE_REQUEST)
+                                            .setService(serviceName)
+                                            .setParams(Utils.objectToByteString(answer))
+                                            .build());
+                                } else {
+                                    finishedLatch.countDown();
+                                    throwables[0] = new Throwable("type must be 'response'");
+                                }
+                            }
+                        });
+                    } catch (Exception e) {
+                        System.out.println(e.getMessage());
+                        sendStreamRef.get().onNext(FunctionsPayload.newBuilder()
+                                .setId(payload.getId())
+                                .setType(Constants.TYPE_SERVICE_REQUEST)
+                                .setService(serviceName)
+                                .setError(e.getMessage())
+                                .build());
+                    }
+                } else {
+                    sendStreamRef.get().onNext(FunctionsPayload.newBuilder()
+                            .setId(payload.getId())
+                            .setType(Constants.TYPE_SERVICE_REQUEST)
+                            .setService(serviceName)
+                            .setError("Function not Registered")
+                            .build());
+                }
+            }
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            throwables[0] = new Throwable(RETRY);
+            finishedLatch.countDown();
+        }
+
+        @Override
+        public void onCompleted() {
+            finishedLatch.countDown();
+        }
+    };
+
+    public Service(Config config, String serviceName) {
+        this.config = config;
+        this.serviceName = serviceName;
+        this.uuid = UUID.randomUUID().toString();
+    }
+
+    public void registerFunc(String functionName, ServiceFunction serviceFunction) {
+        methods.put(functionName, serviceFunction);
+    }
+
+    public void start() throws Throwable {
+        while(true) {
```

Comment:
You are trying to start the stream indefinitely. Shouldn't you attempt to register when the state of the channel is ready?